src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+    Assignment, Literal, Node, Schedule, Loop
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.transformations.transformation_error \
+    import TransformationError
+
+# pylint: disable=too-many-locals
+# pylint: disable=too-many-branches
+
+
+class NemoArrayAccess2LoopTrans(Transformation):
+    '''Provides a transformation from a PSyIR ArrayReference access to a
+    PSyIR NemoLoop. For example:
+
+    >>> from psyclone.parse.algorithm import parse
+    >>> from psyclone.psyGen import PSyFactory
+    >>> from psyclone.domain.nemo.transformations import \
+            NemoArrayAccess2LoopTrans
+    >>> from psyclone.psyir.nodes import Assignment
+    >>> from psyclone.transformations import TransformationError
+    >>>
+    >>> api = "nemo"
+    >>> filename = "tra_adv.F90" # examples/nemo/code
+    >>> ast, invoke_info = parse(filename, api=api)
+    >>> psy = PSyFactory(api).create(invoke_info)
+    >>> schedule = psy.invokes.invoke_list[0].schedule
+    >>> schedule.view()
+    >>>
+    >>> trans = NemoArrayAccess2LoopTrans()
+    >>> for assignment in schedule.walk(Assignment):
+    >>>     for index in asignment.lhs.children:
+    >>>         try:
+    >>>             trans.apply(index)
+    >>>         except TransformationError:
+    >>>             pass
+    >>> schedule.view()
+
+    '''
+    def apply(self, node, options=None):
+        '''Apply the NemoArrayAccess2Loop transformation if the supplied node
+        is an access to an array index within an Array Reference that
+        is on the left-hand-side of an Assignment node. The access
+        must be a scalar (i.e. not a range) and must not include a
+        loop variable (as we are transforming a single access to a
+        loop).
+
+        These constraints are required for correctness and an
+        exception will be raised if they are not satisfied. If the
+        constraints are satisfied then the array access is replaced
+        with a loop iterator and a single trip loop.
+
+        If there are existing loops then the new loop will be placed
+        at the expected nesting, complying to the ordering expected by
+        the NEMO API (specified in the config file). If the existing
+        loops do not conform to this expected ordering then an
+        exception will be raised.
+
+        The name of the loop index is taken from the PSyclone
+        configuration file if a name exists for the particular array
+        index, otherwise a new name is generated.
+
+        :param node: an array index.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        self.validate(node)
+
+        array_index = node.position
```

Comment:
I don't think this will work if `node` is a member of a structure. I think you need `node.parent.indices.index(node)`.

Reply:
The transformation currently raises an exception in the validation if the array is a member of a structure. I don't need this functionality at the moment so I've simply added a test that confirms that an exception is raised in this case.